zip-0308.rst

```diff
@@ -0,0 +1,326 @@
+privacy.
+
+
+Specification
+=============
+
+There are two main aspects to a strategy for selecting migration transactions:
+
+* how many transactions are sent, and when;
+* the amount sent in each transaction.
+
+
+Transaction schedule
+--------------------
+
+When migration is enabled, a node will send up to 5 transactions for inclusion
+in each block with height a multiple of 500 (that is, they are sent immediately
+after seeing a block with height 499, modulo 500).
+
+The migration transactions to be sent in a particular batch can take
+significant time to generate, and this time depends on the speed of the user's
+computer. If they were generated only after a block is seen at the target
+height minus 1, then this could leak information. Therefore, for target
+height N, implementations SHOULD start generating the transactions at around
+height N-5. Each migration transaction SHOULD specify an anchor at height N-10
+for each Sprout JoinSplit description.
+
+Open questions:
+
+* does this reliably give sufficient time to generate the transactions?
+* what happens to a batch if the anchor is invalidated -- should it be
+  regenerated, or cancelled?
+
+Rationale
+'''''''''
+
+Privacy is increased when the times at which to send transactions are
+coordinated between nodes. We choose to send a batch of transactions at each
+coordinated time. Sending multiple transactions in each batch ensures that:
+
+* less information about balances is leaked;
+* it is easier to finish in a reasonable length of time.
+
+The choice of 500 blocks as the batch interval ensures that each batch occurs
+at a different time of day, which may help to mitigate problems with the
+availability of nodes being correlated with the local time-of-day.
+
+Simulations show that the migration process will typically complete reasonably
+quickly even if the amount to be migrated is large:
+
++-----------+--------------------------------------------+
+|           |     Time in days to complete migration     |
+|   Amount  +-----------------+--------+-----------------+
+|           | 10th-percentile | median | 90th-percentile |
++===========+=================+========+=================+
+|      1 ZEC|       1.01      |  1.50  |       2.33      |
++-----------+-----------------+--------+-----------------+
+|     10 ZEC|       1.63      |  2.46  |       3.45      |
++-----------+-----------------+--------+-----------------+
+|    100 ZEC|       2.46      |  3.91  |       5.72      |
++-----------+-----------------+--------+-----------------+
+|   1000 ZEC|       3.74      |  6.01  |       8.83      |
++-----------+-----------------+--------+-----------------+
+|  10000 ZEC|      13.57      | 17.40  |      21.77      |
++-----------+-----------------+--------+-----------------+
+
+The simulation also depends on the amounts sent as specified in the next
+section. It includes the time spent waiting for the first batch to be sent.
+
+
+How much to send in each transaction
```

Comment:
Alternative approach: at each sending slot (e.g., multiple-of-500-blocks), send the largest power-of-2 zatoshis that's not larger than the Sprout balance, not smaller than 0.01 ZEC, and not larger than 1000 ZEC.

This trivial algorithm fulfills the given rationale (and moreover makes the Sapling funds available much faster), so additional rational is needed to justify the more complex proposal.

Reply:
A bit better, to avoid potentially-recognizable monotonically-decreasing sequences: at every sending slot, pick a random index i such that the i-th bit in the binary representation of the Sprout balance is 1, and migrate 2^i zashois. (With the obvious adjustments for too-small and too-large amounts.)

Reply:
I think this does leak more information, and the suggested mitigation for the recognizable-sequence problem is as complicated as the original proposal.

Reply:
For one, I'm deeply suspicious of the magic numbers in the proposed algorithm (which is
`draw from {1,...,99}*10^{6,...,9} until it's less than the balance`). Why powers of 10, and why 2 digits in the mantissa?
Let's take it to the binary extreme: `draw from 2^{20,...36} until it's less than the balance`.
That makes reasoning easier, and strengthens the knapsack rationale. It also avoids the direct leakage of bit value from my 2nd proposal.

Reply:
Powers of 10 means that there is never any small change in the drawn amount; the potential random coins that can be drawn have a floor of 0.01 ZEC. Users use their funds in base 10, not base 2, so the distribution of the resulting Sapling notes is closer to what a user might actually have created, and thus less identifying information might be leaked down the line when those post-migration notes themselves are spent.

As for the two-digit mantissa, it makes the random coins at the edge of the distribution less likely than those in the middle; I'm not sure how I feel about that over a purely-flat distribution.

I do agree that a simpler algorithm here would be nice, but I'm currently more in favour of generating base-10 random coins than base-2.